### ISSUE TITLE
Drop `doctests` from 32-bit CI testing

### ISSUE THREAD

**Issue by WilliamJamieson** [astropy/astropy#15044]
Note that this is a follow up to #15007

It was noted in today's (2023-07-07) dev telecon that one of main problems raised by the 32-bit CI mainly concerned different numbers of digits being displayed for floating point numbers, particularly with tables. It was further noted that often fixes for these errors caused issues for the windows job, which is arguably more important. Moreover, our published documentation is all built on 64-bit linux machines, so the documentation being "completely correct" in the 32-bit setting has less utility especially considering most of the time it requires us to take extra steps to make sure the floats all display the same on different systems for the documentation examples.

Thus I propose to turn off the `doctests` for the 32-bit CI job because the `doctests` ultimately have limited utility in the 32-bit setting, and it will limit the "noise" of inconsequential failures that the infrastructure team needs to deal with after each scheduled Weekly Cron.

**Comment by eerovaher** [comment]
When this was brought up during the dev telecon I remembered running into some problems with the 32-bit CI when I was updating `astropy.table` documentation in #12011, but because that pull request is two years old I didn't remember what the problems were exactly. Having refreshed my memory by looking at the comments there I can comment the current issue better.

Doctests with `Table` instances with `float` columns can indeed be problematic ([see an example together with the explanation](https://github.com/astropy/astropy/pull/12011#discussion_r682524353)), but that doesn't have anything to do with the 32-bit platform. It is nonetheless true that [at least one doctest is being skipped because of the 32-bit CI job](https://github.com/astropy/astropy/pull/12011#issuecomment-892260177). If we wouldn't have to worry about the 32-bit doctests then we could skip the aforementioned test on Windows only, instead of skipping it unconditionally like we are doing now.

Another problem for `astropy.table` doctests are examples of tables that contain integer columns and where the `dtype` of the column is visible, e.g.:
https://github.com/astropy/astropy/blob/015bb1c0eeb9118cef5e33fcfa8f190a2e1281df/docs/table/index.rst?plain=1#L82-L86
On most platforms used in the CI jobs the default integer `dtype` in `numpy` is `int64`, but on Windows and the 32-bit platform it is `int32`. The way `astropy.table` documentation handles this discrepancy is by explicitly specifying the `dtype` for integers instead of using the default, e.g.:
https://github.com/astropy/astropy/blob/015bb1c0eeb9118cef5e33fcfa8f190a2e1281df/docs/table/index.rst?plain=1#L45
The problem with this approach is that it makes the tests more verbose and might create a false impression that using `Table` is more cumbersome than it actually is. If the 32-bit CI job would not include doctests then it would be possible to simplify a lot of that doctest setup at the cost of skipping those doctests on Windows, but in my opinion this would be a worthy tradeoff.

#### TLDR

If we drop doctests from the 32-bit CI job then we could test more code examples and could simplify setup for doctests involving tables with integer columns.

**Comment by pllim** [comment]
I just want to put it on the record that during the telecon, there was also suggestion to drop doctest on Windows but I opposed that one.

**Comment by WilliamJamieson** [comment]
I agree that the windows doctests should stay, we have several windows users/developers who do run the tests. It would be good to keep the doctests functional for them. However, I doubt we have anyone seriously developing on a 32-bit system anymore. I can see astropy being used on older raspberry pi systems or legacy hardware, but I doubt that there is any significant development going on in those cases.

**Comment by mhvk** [comment]
For doctests, I agree it is fine to drop some platforms. Indeed, I think it is fine to ensure they just work on latest stable python and numpy -- they should not get in the way of the clarity of the documentation. So, definitely :+1: to dropping it on 32 bit. For windows, can we run on a version that has 64 bit int by default, or does that still not exist? Though I think this may soon be moot anyway -- I recall numpy is standardizing the default int.

**Comment by pllim** [comment]
This would naturally close out #5744

**Comment by taldcroft** [comment]
I'm 👍  on this.

**Comment by hamogu** [comment]
I'm fine with dropping doctests on 32 bit.

**Comment by pllim** [comment]
Given how @WilliamJamieson refactored 32-bit testing, I think if we go ahead with this, we now also drop doctest on all the "exotic" archs?

**Comment by pllim** [comment]
At the dev telecon on 2023-09-01, those attended agreed that we should move forward with this. But I'll ping @olebole just in case. Keep in mind that only affects doctest, not the whole test suite.

**Comment by olebole** [comment]
As I anyway tend to exclude tests that are not working on some special architectures, I am in principle fine with this. Tests in Debian are mainly to ensure the whole environment (dependencies) work well.
However, I would appreciate if the exclusion could be done within astropy itself, and not in the Github actions: Calling pytest differently depending on 32/64 bit is hard in our test environment. Would that be possible?

**Comment by pllim** [comment]
I thought about this and I am actually not sure how to accomplish this from "within astropy itself". The easiest solution is to selectively call pytest with the `--doctest-plus` flag when we want doctest. Though I think we still have to install `pytest-doctestplus` regardless due to all the custom doctestplus flags we have sprinkled in the test suite (if not, pytest would complain about unrecognized options?).

https://github.com/scientific-python/pytest-doctestplus/blob/c93a23110324d367713dbbbfbcb9d6d32ed74a36/pytest_doctestplus/plugin.py#L87
